<h1>Data cleaning with pandas main operations</h1>


In [2]:
import pandas as pd
import numpy as np

In [25]:
df = pd.read_csv('BL-Flickr-Images-Book.csv')
df.head()
df.shape

(8287, 15)

Dropping columns

In [27]:
to_drop = ['Edition Statement',
               'Corporate Author',
               'Corporate Contributors',
               'Former owner',
               'Engraver',
               'Contributors',
               'Issuance type',
               'Shelfmarks']

In [28]:
df.drop(to_drop, inplace=True, axis=1)

In [29]:
df.shape

(8287, 7)

Changing the Index of a DataFrame

In [30]:
df['Identifier'].is_unique

True

In [32]:
# replace the existing index with this column using set_index
df = df.set_index('Identifier')
# or
# df.set_index('Identifier', inplace=True)

Access each record in a straightforward way with loc[]

In [31]:
df.loc[206]

Identifier                                                          77554
Place of Publication                                               London
Date of Publication                                                  1848
Publisher                                                 Richard Bentley
Title                   Rambles in the romantic regions of the Hartz M...
Author                                   Andersen, H. C. (Hans Christian)
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 206, dtype: object

In [34]:
df.dtypes.value_counts()

object    6
dtype: int64

Let's have a look at date of publication. We may need ot do some data cleaning with dates

In [35]:
df.loc[1905:, 'Date of Publication'].head(10)

Identifier
1905           1888
1929    1839, 38-54
2836           1897
2854           1865
2956        1860-63
2957           1873
3017           1866
3131           1899
4598           1814
4884           1820
Name: Date of Publication, dtype: object

Indeed. What we need to do is:
- Remove the extra dates in square brackets, wherever present: 1879 [1878]
- Convert date ranges to their “start date”, wherever present: 1860-63; 1839, 38-54
- Completely remove the dates we are not certain about and replace them with NumPy’s NaN: [1897?]
- Convert the string nan to NumPy’s NaN value


In [38]:
extr = df['Date of Publication'].str.extract(r'^(\d{4})', expand=False)
extr.head()

Identifier
206    1879
216    1868
218    1869
472    1851
480    1857
Name: Date of Publication, dtype: object

In [42]:
df['Date of Publication'].dtype
df['Date of Publication'] = pd.to_numeric(extr)
df['Date of Publication'].dtype

dtype('float64')

In [44]:
df['Date of Publication'].isnull().sum() / len(df)

0.11717147339205986

Cleaning columns using the .apply function

In [52]:
unwanted_characters = ['[', ',', '-']

def clean_dates(item):
    dop= str(item.loc['Date of Publication'])
    
    if dop == 'nan' or dop[0] == '[':
        return np.NaN
    
    for character in unwanted_characters:
        if character in dop:
            character_index = dop.find(character)
            dop = dop[:character_index]
    
    return dop

df['Date of Publication'] = df.apply(clean_dates, axis = 1)


In [53]:
df.head()

Place of Publication Date of Publication              Publisher  \
Identifier                                                                   
206                      London              1879.0       S. Tinsley & Co.   
216                      London              1868.0           Virtue & Co.   
218                      London              1869.0  Bradbury, Evans & Co.   
472                      London              1851.0          James Darling   
480                      London              1857.0   Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

Let's have a look at place of publication. We may need ot do some data cleaning with place and np.where

In [45]:
df['Place of Publication'].head(10)

Identifier
206                                  London
216                London; Virtue & Yorston
218                                  London
472                                  London
480                                  London
481                                  London
519                                  London
667     pp. 40. G. Bryan & Co: Oxford, 1898
874                                 London]
1143                                 London
Name: Place of Publication, dtype: object

In [49]:
pub = df['Place of Publication']
london = pub.str.contains('London')
london[:5]
oxford = pub.str.contains('Oxford')

In [50]:
df['Place of Publication'] = np.where(london, 'London',
                                      np.where(oxford, 'Oxford',
                                               pub.str.replace('-', ' ')))

df['Place of Publication'].head()

Identifier
206    London
216    London
218    London
472    London
480    London
Name: Place of Publication, dtype: object

An alternative way to do that could be

In [54]:
pub = df['Place of Publication']
df['Place of Publication'] = np.where(pub.str.contains('London'), 'London',
    np.where(pub.str.contains('Oxford'), 'Oxford',
        np.where(pub.eq('Newcastle upon Tyne'),
            'Newcastle-upon-Tyne', df['Place of Publication'])))

In [56]:
df.tail()

Place of Publication Date of Publication                Publisher  \
Identifier                                                                     
4158088                  London              1838.0                      NaN   
4158128                   Derby              1831.0          M. Mozley & Son   
4159563                  London                 NaN  T. Cadell and W. Davies   
4159587     Newcastle-upon-Tyne              1834.0         Mackenzie & Dent   
4160339                  London              1834.0                      NaN   

                                                        Title  \
Identifier                                                      
4158088     The Parochial History of Cornwall, founded on,...   
4158128     The History and Gazetteer of the County of Der...   
4159563     Magna Britannia; being a concise topographical...   
4159587     An historical, topographical and descriptive v...   
4160339     Collectanea Topographica et Genealogica. [Firs...   

                                                       Author  \
Identifier                                                      
4158088                    GIDDY, afterwards GILBERT, Davies.   
4158128                            GLOVER, Stephen - of Derby   
4159563     LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
4159587                                 Mackenzie, E. (Eneas)   
4160339                                                   NaN   

                                                   Flickr URL  
Identifier                                                     
4158088     http://www.flickr.com/photos/britishlibrary/ta...  
4158128     http://www.flickr.com/photos/britishlibrary/ta...  
4159563     http://www.flickr.com/photos/britishlibrary/ta...  
4159587     http://www.flickr.com/photos/britishlibrary/ta...  
4160339     http://www.flickr.com/photos/britishlibrary/ta...

Cleaning the Entire Dataset Using the applymap Function

In [58]:

university_towns = []

with open('university_towns.txt', 'r') as file:
    items = file.readlines()
    states = list(filter(lambda x: '[edit]' in x, items))
    
    for index, state in enumerate(states):
        start = items.index(state) + 1
        if index == 49: #since 50 states
            end = len(items)
        else:
            end = items.index(states[index + 1])
            
        pairs = map(lambda x: [state, x], items[start:end])
        university_towns.extend(pairs)
        
towns_df = pd.DataFrame(university_towns, columns = ['State', 'RegionName'])
towns_df.head()

State                                         RegionName
0  Alabama[edit]\n                    Auburn (Auburn University)[1]\n
1  Alabama[edit]\n           Florence (University of North Alabama)\n
2  Alabama[edit]\n  Jacksonville (Jacksonville State University)[2]\n
3  Alabama[edit]\n       Livingston (University of West Alabama)[2]\n
4  Alabama[edit]\n         Montevallo (University of Montevallo)[2]\n

In [59]:
towns_df = pd.DataFrame(university_towns,  columns=['State', 'RegionName'])

def clean_up(item):
    if '(' in item:
        return item[:item.find('(') - 1]
    
    if '[' in item:
        return item[:item.find('[')]
    

towns_df =  towns_df.applymap(clean_up)
towns_df.head()


State    RegionName
0  Alabama        Auburn
1  Alabama      Florence
2  Alabama  Jacksonville
3  Alabama    Livingston
4  Alabama    Montevallo